api/src/main/java/com/cloud/agent/api/storage/OVFHelper.java

```diff
@@ -366,21 +387,25 @@ private DatadiskTO generateDiskTO(OVFFile file, OVFDisk disk, String ovfParentPa
                 isIso, bootable, controller, controllerSubType, diskNumber, configuration);
     }
 
-    protected List<OVFDisk> extractDisksFromOvfDocumentTree(Document doc) {
-        NodeList disks = doc.getElementsByTagName("Disk");
-        NodeList ovfDisks = doc.getElementsByTagName("ovf:Disk");
-        NodeList items = doc.getElementsByTagName("Item");
+    protected NodeList getElementByTagNameAndSupportedNamespaces(Document doc, List<String> namespaces, String tagName) {
+        NodeList nodeList = null;
+        for (String namespace : namespaces) {
+            nodeList = doc.getElementsByTagNameNS(namespace, tagName);
+            if (nodeList.getLength() > 0) {
+                break;
+            }
+        }
+        return nodeList;
+    }
 
-        int totalDisksLength = disks.getLength() + ovfDisks.getLength();
+    protected List<OVFDisk> extractDisksFromOvfDocumentTree(Document doc, List<String> supportedNamespaces) {
+        NodeList disks = getElementByTagNameAndSupportedNamespaces(doc, supportedNamespaces, "Disk");
```

Comment:
this will work in most cases. the only quirk is if there are two name spaces defined that both have a `Disk` element.
Can we specify the exact name space we need?

Reply:
Yes we could, but in that case which namespace should be used? If both `Disk` elements have valid prefixes/namespaces then I think the method should retrieve all the elements defined in the file even with different namespaces, do you agree?

Reply:
No, they would have different meaning. Only the one that is meant to be in the name space that usually uses `ovf`as prefix is an actual disk template. I am not to worried about this occurring with high probability but it is a real option for xml-composers.